### PR TITLE
Patched class loader and pluglinlib

### PR DESCRIPTION
**Pull request by ivanpauno**
Parched pluginlib and class_loader in order to remove poco.

I commented out almost all library load related functions or emptied them, as they were doing nothing.
The register class and create_instance parts of the code should work as before.
Didn't tested on run time.

Maybe, this is more complicated to maintain that building poco.

### CODE REVIEW

File: patches/class_loader.patch [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+ #include "class_loader/meta_object.hpp"
+@@ -64,8 +66,12 @@ typedef std::string ClassName;
+ typedef std::string BaseClassName;
+ typedef std::map<ClassName, class_loader_private::AbstractMetaObjectBase *> FactoryMap;
+ typedef std::map<BaseClassName, FactoryMap> BaseToFactoryMapMap;
++#ifndef ANDROID
+ typedef std::pair<LibraryPath, Poco::SharedLibrary *> LibraryPair;
+ typedef std::vector<LibraryPair> LibraryVector;
++// #else
++// typedef std::vector<int> LibraryVector;
```
Comment (jubeira) [inline patches/class_loader.patch:75]: Alright, so this means we can just remove this typedef in the end?

File: patches/class_loader.patch [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+ {
++#ifndef ANDROID
+   printf("*******************************************************************************\n");
+   printf("*****               class_loader_private DEBUG INFORMATION                   *****\n");
+   printf("*******************************************************************************\n");
+@@ -646,6 +659,7 @@ void printDebugInfoToScreen()
+ 
+   printf("********************************** END DEBUG **********************************\n");
+   printf("*******************************************************************************\n\n");
++#endif
```
Comment (jubeira) [inline patches/class_loader.patch:205]: Are we leaving something here? Perhaps one big `ifndef` is enough?

File: patches/pluginlib.patch [hunk 3]
```diff
@@ -12,31 +30,87 @@
...
+     std::string class_type = getClassType(lookup_name);
+@@ -225,9 +243,11 @@ T * ClassLoader<T>::createUnmanagedInstance(const std::string & lookup_name)
+   ROS_DEBUG_NAMED("pluginlib.ClassLoader", "Attempting to create UNMANAGED instance for class %s.",
+     lookup_name.c_str());
+ 
++#ifndef ANDROID
+   if (!isClassLoaded(lookup_name)) {
+     loadLibraryForClass(lookup_name);
+   }
++#endif
```
Comment (jubeira) [inline patches/pluginlib.patch:100]: Hmm what if we patch `isClassLoaded` to return `true` always instead of adding so many `ifndef`?

**Review by jubeira — COMMENTED** [review]
I did a high level pass and made some comments; I haven't tried a fresh build. 

Good job so far!

File: patches/class_loader.patch (hunk 1 above)
Comment (ivanpauno) [reply patches/class_loader.patch:75]: Yes, I commented out the functions that used it in the prototype, so yes.
I forgot deleting the comment.

File: patches/pluginlib.patch (hunk 3 above)
Comment (ivanpauno) [reply patches/pluginlib.patch:100]: Maybe. Also, we could remove the old patch inside loadLibraryForClass().

**Comment by meyerj** [comment]
Added a patch to actually apply `class_loader.patch`.

File: do_everything.sh [hunk 4]
```diff
@@ -273,13 +272,14 @@ if [[ $skip -ne 1 ]] ; then
     # Wait for next release to remove (current 1.12.4)
     # apply_patch $my_loc/patches/global_planner.patch
 
-    #Patch Poco lib
-    apply_patch $my_loc/patches/poco.patch
```
Comment (meyerj) [inline do_everything.sh:277]: remove `poco.patch`

**Review by meyerj — COMMENTED** [review]
I tested this patch and it compiles fine (with https://github.com/Intermodalics/ros_android/pull/62/commits/5ba9639439f6945bccdf1fecfe05d0a16e7edf92), but although I linked our application to `libimage_transport_plugins.a` it emits the following error at startup:
```
terminating with uncaught exception of type image_transport::TransportLoadException: Unable to load plugin for transport 'raw', error string:
MultiLibraryClassLoader: Could not create object of class type image_transport::RawSubscriber as no factory exists for it. Make sure that the library exists and was explicitly loaded through MultiLibraryClassLoader::loadLibrary()
Aborted 
```

```cmake
# CMakeLists.txt
find_library(image_transport_plugins_LIBRARY NAMES image_transport_plugins)
target_link_libraries(${PROJECT_NAME}_app
  -Wl,--whole-archive ${image_transport_plugins_LIBRARY} -Wl,--no-whole-archive
)
```

This approach did work before. I am not sure whether `-Wl,--whole-archive` is actually required, but it is the equivalent of `LOCAL_WHOLE_STATIC_LIBRARIES` in `Android.mk` as described [here](http://wiki.ros.org/android_ndk/Tutorials/UsingPluginlib).

In this special case we should link `libimage_transport_plugins.a` by default in package `image_transport`, such that at least raw publishing is available without any additional CMake commands.

**Comment by meyerj** [comment]
I think we can close this for now. Indeed the overhead for maintaining this patch might be higher than simply building with poco. Anyway, thanks for looking into this, @ivanpauno.

On the contrary, what is actually the fundamental problem that makes it necessary to discover plugins at build-time and link to static libraries? In principle it should be possible to use [dlopen()](https://developer.android.com/ndk/guides/stable_apis#dynamic-linker-lib) on Android, too. Apparently there also have been [many changes introduced with API level 23 or 24](https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md), which might have solved some issues with earlier versions. Sure, this would require a lot of patches to add the `SHARED` keyword explicitly to all `add_library()` calls for plugin libraries, and the required plugin libs would need to be packaged into the apk somehow to make them available on Android. Also the package and nodelet manifests would need to be accessible in a simulated install-space pointed to by `ROS_PACKAGE_PATH` such that plugin discovery works in the same way as on Linux.

Just thinking out loud...

**Pull request closed without merging**